draft-how-does-mutant-work.markdown

```diff
@@ -0,0 +1,431 @@
+
+```Ruby
+reporter:          Reporter::CLI.build($stdout)
+```
+
+The default reporter. Keep that in mind, we'll come back to this later.
+
+```Ruby
+load_path:         $LOAD_PATH,
+requires:          EMPTY_ARRAY,
+pathname:          Pathname,
+```
+
+What I like here is how configurable `Mutant`. You can even inject another $LOAD_PATH, additional requires or another Pathname'esque library.
+
+**The big picture for now:**
+
+TODO: proper picture
+
+bin -> cli -> config
+
+### Environment
+
+So where were we?
+Keep in mind that we are still talking about this:
+
+```Ruby
+Runner.call(
+  Env::Bootstrap.call(
+    call(arguments) # -> This part we just discussed.
+  )
+).success?
+```
+
+Now that we know what 
+
+```Ruby
+call(arguments) # -> This part we just discussed.
+```
+
+does, let's look at:
+
+```Ruby
+Env::Bootstrap.call(
+  call(arguments) # -> This part we just discussed.
+)
+```
+
+Again, this is `Procto` at work.
+
+```Ruby
+Env::Bootstrap.call configuration
+```
+
+means
+
+```Ruby
+ # "some_method" is what we configure when including Procto
+Env::Bootstrap.new(configuration).some_method
+```
+
+So let's check out the class declaration of Env::Bootstrap with a focus on the initializer:
+
+```Ruby
+module Mutant
+  class Env
+    class Bootstrap
+      include Procto.call(:env) # -> That's the "some_method" from above
+
+      def initialize(*)
+        super
+        @parser = Parser.new
+        infect
+        initialize_matchable_scopes
+      end
+    end
+  end
+end
+```
+
+Step by step:
+
+* `super`: TODO: Explain Concord?
+* `Parser.new`: `Mutant` uses the awesome `[parser](https://github.com/whitequark/parser)` gem for producing ASTs from source code and has its own thin wrapper around
+* `infect`: TODO
+* `initialize_matchable_scopes`: TODO
+
+Now that we know how the initializer works we know half of the story of the call to 
+
+```Ruby
+Runner.call(
+  Env::Bootstrap.call(
+    call(arguments)
+  )
+).success?
+```
+
+But what happens when we call `Env::Bootstrap.call`? If you remember the `include` statement in the class declaration:
+
+```Ruby
+module Mutant
+  class Env
+    class Bootstrap
+      include Procto.call(:env)
+
+      def initialize(*)
+        # snip, we just discussed this
+      end
+    end
+  end
+end
+```
+
+you can see that we actually call `env` on our `Env::Bootstrap` object:
+
+```Ruby
+  def env
+    subjects = matched_subjects
+    Env.new(
+      actor_env:        Actor::Env.new(Thread),
+      config:           config,
+      integration:      @integration,
+      matchable_scopes: matchable_scopes,
+      mutations:        subjects.flat_map(&:mutations),
+      parser:           parser,
+      selector:         Selector::Expression.new(@integration),
+      subjects:         subjects
+    )
+  end
+```
+
+TODO: Explain that :)
+
+```
+TODO: Debug your way through in cli.rb via pry
+args = call(arguments); nil
+env = Env::Bootstrap.new args
+cd env
+wtf are matched_subjects?
```

Comment:
The ones that end up in the selected subjects for mutation testing, explained here: https://github.com/troessner/articles/pull/3/files#r67711942